### PR TITLE
Adding FMA folding in the VM.

### PR DESCRIPTION
**Pull request by benvanik**
benchmarks: all

### CODE REVIEW

**Comment by pzread** [comment]
Noticed the typo in the benchmark trailer, should be `benchmarks: all` instead of `benchmark: all` (I know this is not good... fixing on the way)

**Comment by benvanik** [comment]
ack, thanks - I was sitting and waiting to see if they'd run and would have been waiting awhile :)

**Comment by benvanik** [comment]
(how do I get it to run again now?)

**Comment by pzread** [comment]
> (how do I get it to run again now?)

The current way is still `git commit --allow-empty -m 'trigger build'; git push` to trigger the build again (or you have a new commit to push)... I'll probably send a patch first to fetch the trailer when the workflow starts, at least re-run the workflow will fetch the new trailer (currently Github sends the cached PR description when re-run is triggered)

**Comment by benvanik** [comment]
(tests seem ok but summary failed to post or something)

File: compiler/src/iree/compiler/Dialect/VM/IR/VMOpFolders.cpp [hunk 1]
```diff
@@ -607,6 +607,38 @@ static Attribute constFoldTernaryOp(Attribute rawA, Attribute rawB,
...
+                             a.getType(), a, b, c)
+              .getResult());
+      return success();
+    };
+    if (auto mulOp = dyn_cast_or_null<MulOp>(addOp.getLhs().getDefiningOp())) {
+      return fuse(mulOp, mulOp.getLhs(), mulOp.getRhs(), addOp.getRhs());
+    } else if (auto mulOp =
+                   dyn_cast_or_null<MulOp>(addOp.getRhs().getDefiningOp())) {
+      return fuse(mulOp, mulOp.getLhs(), mulOp.getRhs(), addOp.getLhs());
+    }
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/VM/IR/VMOpFolders.cpp:637]: Worth testing LHS and RHS handling?

File: compiler/src/iree/compiler/Dialect/VM/IR/test/arithmetic_folding.mlir [hunk 2]
```diff
@@ -29,6 +29,46 @@ vm.module @add_i32_folds {
...
   }
+
+  // CHECK-LABEL: @mul_add_i32
+  vm.func @mul_add_i32(%arg0: i32, %arg1: i32, %arg2: i32) -> i32 {
+    %0 = vm.mul.i32 %arg0, %arg1 : i32
+    // CHECK: %[[RET:.+]] = vm.fma.i32 %arg0, %arg1, %arg2 : i32
+    %1 = vm.add.i32 %0, %arg2 : i32
+    // CHECK: return %[[RET]]
+    vm.return %1 : i32
+  }
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Dialect/VM/IR/test/arithmetic_folding.mlir]: (could expand this to cover LHS and RHS)

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**